tests/integration/ha/test_horizontal_scaling.py

```diff
@@ -74,3 +86,141 @@ async def test_horizontal_scale_up(ops_test: OpsTest) -> None:
+
+
+@pytest.mark.ha_tests
+@pytest.mark.abort_on_fail
+async def test_safe_scale_down_shards_realloc(ops_test: OpsTest) -> None:
+    """Tests the shutdown of a node, and re-allocation of shards to a newly joined unit.
+
+    The goal of this test is to make sure that shards are automatically relocated after
+    a Yellow status on the cluster caused by a scale-down event.
+    """
+    # scale up
+    await ops_test.model.applications[APP_NAME].add_unit(count=1)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", timeout=1000, wait_for_exact_units=4
+    )
+
+    leader_unit_ip = await get_leader_unit_ip(ops_test)
+    leader_unit_id = await get_leader_unit_id(ops_test)
+
+    # fetch all nodes
+    unit_ids = get_application_unit_ids(ops_test)
+    unit_id_to_stop = [unit_id for unit_id in unit_ids if unit_id != leader_unit_id][0]
+    unit_ids_to_keep = [unit_id for unit_id in unit_ids if unit_id != unit_id_to_stop]
+
+    # create indices with right num of primary and replica shards, and populate with data
+    await create_dummy_indexes(ops_test, leader_unit_ip)
+    await create_dummy_docs(ops_test, leader_unit_ip)
+
+    # get initial cluster health - expected to be all good: green
+    cluster_health_resp = await cluster_health(ops_test, leader_unit_ip)
+    assert cluster_health_resp["status"] == "green"
+    assert cluster_health_resp["unassigned_shards"] == 0
+
+    # get initial cluster allocation (nodes and their corresponding shards)
+    init_shards_per_node = await get_number_of_shards_by_node(ops_test, leader_unit_ip)
+    assert init_shards_per_node.get(-1, 0) == 0  # unallocated shards
+    print(await cluster_allocation(ops_test, leader_unit_ip))
+
+    # remove the service in the chosen unit
+    # await run_action(ops_test, unit_id_to_stop, "stop-service")
+    await ops_test.model.applications[APP_NAME].destroy_unit(f"{APP_NAME}/{unit_id_to_stop}")
+    # await ops_test.model.applications[APP_NAME].add_unit(count=-1)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", timeout=1000, wait_for_exact_units=3
+    )
+
+    # check if at least partial shard re-allocation happened
+    new_shards_per_node = await get_number_of_shards_by_node(ops_test, leader_unit_ip)
+
+    # some shards should have been reallocated, NOT ALL due to already existing replicas elsewhere
+    assert new_shards_per_node.get(-1, 0) > 0  # some shards not reallocated
+
+    are_some_shards_reallocated = False
+    for unit_id in unit_ids_to_keep:
+        are_some_shards_reallocated = (
+            are_some_shards_reallocated
+            or new_shards_per_node[unit_id] > init_shards_per_node[unit_id]
+        )
+    assert are_some_shards_reallocated
+
+    # get new cluster health
+    cluster_health_resp = await cluster_health(ops_test, leader_unit_ip)
+
+    # not all replica shards should have been reallocated
+    assert cluster_health_resp["status"] == "yellow"
```

Comment:
If the cluster wasn't scaled back up, would this ever become green by itself?

Reply:
In this case, it wouldn't. Because the shards settings conditions are not respected, i.e: 2 primary shards + 3 replicas of a each shard, so we at least need 4 nodes in this specific case for the cluster to become green.
Yellow: some replica shards failed to be allocated
Red: some primary shards failed to be allocated

Reply:
OK, thanks for explaining